wsmaster/che-core-api-machine/src/main/java/org/eclipse/che/api/machine/server/model/impl/SnapshotImpl.java

```diff
@@ -13,40 +13,91 @@
 import org.eclipse.che.api.core.model.machine.MachineConfig;
 import org.eclipse.che.api.core.model.machine.MachineSource;
 import org.eclipse.che.api.core.model.machine.Snapshot;
+import org.eclipse.che.api.machine.server.spi.Instance;
 import org.eclipse.che.commons.lang.NameGenerator;
 
+import javax.persistence.Basic;
+import javax.persistence.Embedded;
+import javax.persistence.Entity;
+import javax.persistence.Id;
+import javax.persistence.Index;
+import javax.persistence.NamedQueries;
+import javax.persistence.NamedQuery;
+import javax.persistence.Table;
 import java.util.Objects;
 
-import static java.util.Objects.requireNonNull;
-
 /**
- * Saved state of {@link org.eclipse.che.api.machine.server.spi.Instance}.
+ * Saved state of {@link Instance}.
  *
  * @author Yevhenii Voevodin
  */
+@Entity(name = "Snapshot")
+@NamedQueries(
+        {
+                @NamedQuery(name = "Snapshot.getByMachine",
+                            query = "SELECT snapshot " +
+                                    "FROM Snapshot snapshot " +
+                                    "WHERE snapshot.workspaceId = :workspaceId" +
+                                    "  AND snapshot.envName     = :envName" +
+                                    "  AND snapshot.machineName = :machineName"),
+                @NamedQuery(name = "Snapshot.findSnapshots",
+                            query = "SELECT snapshot " +
```

Comment:
It's [okay](https://devcenter.heroku.com/articles/postgresql-indexes#multi-column-indexes)